File: crates/ivza-core/src/intent/parser.rs
use std::str::FromStr;

use anyhow::{anyhow, Result};
use serde::{Deserialize, Serialize};
use solana_sdk::pubkey::Pubkey;
use tracing::{info, warn};

/// A high-level intent representing what the user wants to accomplish.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Intent {
    /// Type of intent.
    pub intent_type: IntentType,
    /// Parameters for the intent.
    pub params: IntentParams,
    /// Optional user-provided label.
    pub label: Option<String>,
    /// Maximum acceptable slippage in basis points (for swaps).
    pub max_slippage_bps: Option<u64>,
    /// Priority fee the user is willing to pay.
    pub priority_fee: Option<u64>,
}

impl Intent {
    pub fn new(intent_type: IntentType, params: IntentParams) -> Self {
        Self {
            intent_type,
            params,
            label: None,
            max_slippage_bps: None,
            priority_fee: None,
        }
    }

    pub fn with_label(mut self, label: impl Into<String>) -> Self {
        self.label = Some(label.into());
        self
    }

    pub fn with_slippage(mut self, bps: u64) -> Self {
        self.max_slippage_bps = Some(bps);
        self
    }

    pub fn with_priority_fee(mut self, fee: u64) -> Self {
        self.priority_fee = Some(fee);
        self
    }
}

/// The type of intent.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub enum IntentType {
    /// Swap token A for token B.
    Swap,
    /// Multi-hop swap through intermediate tokens.
    MultiHopSwap,
    /// Stake SOL to a validator.
    Stake,
    /// Unstake SOL from a validator.
    Unstake,
    /// Provide liquidity to a pool.
    ProvideLiquidity,
    /// Remove liquidity from a pool.
    RemoveLiquidity,
    /// Transfer tokens.
    Transfer,
    /// Create a token account.
    CreateAccount,
}

/// Parameters for the various intent types.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum IntentParams {
    Swap(SwapParams),
    MultiHopSwap(MultiHopSwapParams),
    Stake(StakeParams),
    Unstake(UnstakeParams),
    ProvideLiquidity(ProvideLiquidityParams),
    RemoveLiquidity(RemoveLiquidityParams),
    Transfer(TransferParams),
    CreateAccount(CreateAccountParams),
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SwapParams {
    pub input_mint: Pubkey,
    pub output_mint: Pubkey,
    pub amount_in: u64,
    pub minimum_amount_out: Option<u64>,
    pub user_wallet: Pubkey,
    pub dex_program: Option<Pubkey>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MultiHopSwapParams {
    pub route: Vec<Pubkey>, // mint addresses: [input, intermediate..., output]
    pub amount_in: u64,
    pub minimum_amount_out: Option<u64>,
    pub user_wallet: Pubkey,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StakeParams {
    pub amount: u64,
    pub validator_vote_account: Pubkey,
    pub user_wallet: Pubkey,
    pub stake_account: Option<Pubkey>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UnstakeParams {
    pub stake_account: Pubkey,
    pub user_wallet: Pubkey,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProvideLiquidityParams {
    pub pool: Pubkey,
    pub token_a_mint: Pubkey,
    pub token_b_mint: Pubkey,
    pub amount_a: u64,
    pub amount_b: u64,
    pub user_wallet: Pubkey,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RemoveLiquidityParams {
    pub pool: Pubkey,
    pub lp_amount: u64,
    pub user_wallet: Pubkey,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TransferParams {
    pub mint: Pubkey,
    pub amount: u64,
    pub from_wallet: Pubkey,
    pub to_wallet: Pubkey,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreateAccountParams {
    pub mint: Pubkey,
    pub owner: Pubkey,
}

/// Parses high-level intent descriptions into structured Intent objects.
///
/// Supports parsing from JSON and from a simple DSL string format.
pub struct IntentParser;

impl IntentParser {
    pub fn new() -> Self {
        Self
    }

    /// Parse an intent from a JSON string.
    pub fn parse_json(&self, json: &str) -> Result<Intent> {
        let raw: serde_json::Value =
            serde_json::from_str(json).map_err(|e| anyhow!("Invalid JSON: {}", e))?;

        let intent_type_str = raw
            .get("type")
            .and_then(|v| v.as_str())
            .ok_or_else(|| anyhow!("Missing 'type' field"))?;

        let intent_type = self.parse_intent_type(intent_type_str)?;

        let params_value = raw
            .get("params")
            .ok_or_else(|| anyhow!("Missing 'params' field"))?;

        let params = self.parse_params(&intent_type, params_value)?;

        let mut intent = Intent::new(intent_type, params);

        if let Some(label) = raw.get("label").and_then(|v| v.as_str()) {
            intent = intent.with_label(label);
        }
        if let Some(slippage) = raw.get("max_slippage_bps").and_then(|v| v.as_u64()) {
            intent = intent.with_slippage(slippage);
        }
        if let Some(fee) = raw.get("priority_fee").and_then(|v| v.as_u64()) {
            intent = intent.with_priority_fee(fee);
        }

        info!("Parsed intent: {:?}", intent.intent_type);
        Ok(intent)
    }

    /// Parse a batch of intents from a JSON array.
    pub fn parse_batch(&self, json: &str) -> Result<Vec<Intent>> {
        let arr: Vec<serde_json::Value> =
            serde_json::from_str(json).map_err(|e| anyhow!("Invalid JSON array: {}", e))?;

        let mut intents = Vec::new();
        for (i, val) in arr.iter().enumerate() {
            let json_str = serde_json::to_string(val)?;
            match self.parse_json(&json_str) {
                Ok(intent) => intents.push(intent),
                Err(e) => {
                    warn!("Failed to parse intent {}: {}", i, e);
                    return Err(anyhow!("Failed to parse intent {}: {}", i, e));
                }
            }
        }
        Ok(intents)
    }

    /// Parse a simple DSL string format.
    /// Format: "swap <amount> <input_mint> for <output_mint> by <wallet>"
    ///         "stake <amount> to <validator> by <wallet>"
    ///         "transfer <amount> <mint> from <wallet> to <wallet>"
    pub fn parse_dsl(&self, input: &str) -> Result<Intent> {
        let tokens: Vec<&str> = input.split_whitespace().collect();
        if tokens.is_empty() {
            return Err(anyhow!("Empty intent string"));
        }

        match tokens[0].to_lowercase().as_str() {
            "swap" => self.parse_swap_dsl(&tokens),
            "stake" => self.parse_stake_dsl(&tokens),
            "unstake" => self.parse_unstake_dsl(&tokens),
            "transfer" => self.parse_transfer_dsl(&tokens),
            "provide-liquidity" | "provide_liquidity" => self.parse_provide_liquidity_dsl(&tokens),
            "remove-liquidity" | "remove_liquidity" => self.parse_remove_liquidity_dsl(&tokens),
            _ => Err(anyhow!("Unknown intent type: {}", tokens[0])),
        }
    }

    fn parse_intent_type(&self, s: &str) -> Result<IntentType> {
        match s.to_lowercase().as_str() {
            "swap" => Ok(IntentType::Swap),
            "multi_hop_swap" | "multihop" | "multi-hop" => Ok(IntentType::MultiHopSwap),
            "stake" => Ok(IntentType::Stake),
            "unstake" => Ok(IntentType::Unstake),
            "provide_liquidity" | "add_liquidity" => Ok(IntentType::ProvideLiquidity),
            "remove_liquidity" => Ok(IntentType::RemoveLiquidity),
            "transfer" => Ok(IntentType::Transfer),
            "create_account" => Ok(IntentType::CreateAccount),
            _ => Err(anyhow!("Unknown intent type: {}", s)),
        }
    }

    fn parse_params(
        &self,
        intent_type: &IntentType,
        value: &serde_json::Value,
    ) -> Result<IntentParams> {
        match intent_type {
            IntentType::Swap => {
                let p: SwapParams = serde_json::from_value(value.clone())?;
                Ok(IntentParams::Swap(p))
            }
            IntentType::MultiHopSwap => {
                let p: MultiHopSwapParams = serde_json::from_value(value.clone())?;
                Ok(IntentParams::MultiHopSwap(p))
            }
            IntentType::Stake => {
                let p: StakeParams = serde_json::from_value(value.clone())?;
                Ok(IntentParams::Stake(p))
            }
            IntentType::Unstake => {
                let p: UnstakeParams = serde_json::from_value(value.clone())?;
                Ok(IntentParams::Unstake(p))
            }
            IntentType::ProvideLiquidity => {
                let p: ProvideLiquidityParams = serde_json::from_value(value.clone())?;
                Ok(IntentParams::ProvideLiquidity(p))
            }
            IntentType::RemoveLiquidity => {
                let p: RemoveLiquidityParams = serde_json::from_value(value.clone())?;
                Ok(IntentParams::RemoveLiquidity(p))
            }
            IntentType::Transfer => {
                let p: TransferParams = serde_json::from_value(value.clone())?;
                Ok(IntentParams::Transfer(p))
            }
            IntentType::CreateAccount => {
                let p: CreateAccountParams = serde_json::from_value(value.clone())?;
                Ok(IntentParams::CreateAccount(p))
            }
        }
    }

    // --- DSL parsers ---

    /// "swap <amount> <input_mint> for <output_mint> by <wallet>"
    fn parse_swap_dsl(&self, tokens: &[&str]) -> Result<Intent> {
        if tokens.len() < 6 {
            return Err(anyhow!(
                "Swap DSL format: swap <amount> <input_mint> for <output_mint> by <wallet>"
            ));
        }
        let amount: u64 = tokens[1].parse().map_err(|_| anyhow!("Invalid amount"))?;
        let input_mint = Pubkey::from_str(tokens[2]).map_err(|_| anyhow!("Invalid input_mint"))?;
        // tokens[3] should be "for"
        let output_mint =
            Pubkey::from_str(tokens[4]).map_err(|_| anyhow!("Invalid output_mint"))?;
        // tokens[5] should be "by"
        let wallet = Pubkey::from_str(tokens[6]).map_err(|_| anyhow!("Invalid wallet"))?;

        Ok(Intent::new(
            IntentType::Swap,
            IntentParams::Swap(SwapParams {
                input_mint,
                output_mint,
                amount_in: amount,
                minimum_amount_out: None,
                user_wallet: wallet,
                dex_program: None,
            }),
        ))
    }

    /// "stake <amount> to <validator> by <wallet>"
    fn parse_stake_dsl(&self, tokens: &[&str]) -> Result<Intent> {
        if tokens.len() < 5 {
            return Err(anyhow!(
                "Stake DSL format: stake <amount> to <validator> by <wallet>"
            ));
        }
        let amount: u64 = tokens[1].parse().map_err(|_| anyhow!("Invalid amount"))?;
        let validator =
            Pubkey::from_str(tokens[3]).map_err(|_| anyhow!("Invalid validator pubkey"))?;
        let wallet = Pubkey::from_str(tokens[5]).map_err(|_| anyhow!("Invalid wallet pubkey"))?;

        Ok(Intent::new(
            IntentType::Stake,
            IntentParams::Stake(StakeParams {
                amount,
                validator_vote_account: validator,
                user_wallet: wallet,
                stake_account: None,
            }),
        ))
    }

    /// "unstake <stake_account> by <wallet>"
    fn parse_unstake_dsl(&self, tokens: &[&str]) -> Result<Intent> {
        if tokens.len() < 4 {
            return Err(anyhow!(
                "Unstake DSL format: unstake <stake_account> by <wallet>"
            ));
        }
        let stake_account =
            Pubkey::from_str(tokens[1]).map_err(|_| anyhow!("Invalid stake account"))?;
        let wallet = Pubkey::from_str(tokens[3]).map_err(|_| anyhow!("Invalid wallet pubkey"))?;

        Ok(Intent::new(
            IntentType::Unstake,
            IntentParams::Unstake(UnstakeParams {
                stake_account,
                user_wallet: wallet,
            }),
        ))
    }

    /// "transfer <amount> <mint> from <wallet> to <wallet>"
    fn parse_transfer_dsl(&self, tokens: &[&str]) -> Result<Intent> {
        if tokens.len() < 7 {
            return Err(anyhow!(
                "Transfer DSL format: transfer <amount> <mint> from <from_wallet> to <to_wallet>"
            ));
        }
        let amount: u64 = tokens[1].parse().map_err(|_| anyhow!("Invalid amount"))?;
        let mint = Pubkey::from_str(tokens[2]).map_err(|_| anyhow!("Invalid mint"))?;
        let from = Pubkey::from_str(tokens[4]).map_err(|_| anyhow!("Invalid from wallet"))?;
        let to = Pubkey::from_str(tokens[6]).map_err(|_| anyhow!("Invalid to wallet"))?;

        Ok(Intent::new(
            IntentType::Transfer,
            IntentParams::Transfer(TransferParams {
                mint,
                amount,
                from_wallet: from,
                to_wallet: to,
            }),
        ))
    }

    /// "provide-liquidity <pool> <amount_a> <mint_a> <amount_b> <mint_b> by <wallet>"
    fn parse_provide_liquidity_dsl(&self, tokens: &[&str]) -> Result<Intent> {
        if tokens.len() < 8 {
            return Err(anyhow!(
                "Provide liquidity DSL format: provide-liquidity <pool> <amount_a> <mint_a> <amount_b> <mint_b> by <wallet>"
            ));
        }
        let pool = Pubkey::from_str(tokens[1]).map_err(|_| anyhow!("Invalid pool"))?;
        let amount_a: u64 = tokens[2].parse().map_err(|_| anyhow!("Invalid amount_a"))?;
        let mint_a = Pubkey::from_str(tokens[3]).map_err(|_| anyhow!("Invalid mint_a"))?;
        let amount_b: u64 = tokens[4].parse().map_err(|_| anyhow!("Invalid amount_b"))?;
        let mint_b = Pubkey::from_str(tokens[5]).map_err(|_| anyhow!("Invalid mint_b"))?;
        let wallet = Pubkey::from_str(tokens[7]).map_err(|_| anyhow!("Invalid wallet"))?;

        Ok(Intent::new(
            IntentType::ProvideLiquidity,
            IntentParams::ProvideLiquidity(ProvideLiquidityParams {
                pool,
                token_a_mint: mint_a,
                token_b_mint: mint_b,
                amount_a,
                amount_b,
                user_wallet: wallet,
            }),
        ))
    }

    /// "remove-liquidity <pool> <lp_amount> by <wallet>"
    fn parse_remove_liquidity_dsl(&self, tokens: &[&str]) -> Result<Intent> {
        if tokens.len() < 5 {
            return Err(anyhow!(
                "Remove liquidity DSL format: remove-liquidity <pool> <lp_amount> by <wallet>"
            ));
        }
        let pool = Pubkey::from_str(tokens[1]).map_err(|_| anyhow!("Invalid pool"))?;
        let lp_amount: u64 = tokens[2]
            .parse()
            .map_err(|_| anyhow!("Invalid lp_amount"))?;
        let wallet = Pubkey::from_str(tokens[4]).map_err(|_| anyhow!("Invalid wallet"))?;

        Ok(Intent::new(
            IntentType::RemoveLiquidity,
            IntentParams::RemoveLiquidity(RemoveLiquidityParams {
                pool,
                lp_amount,
                user_wallet: wallet,
            }),
        ))
    }
}

impl Default for IntentParser {
    fn default() -> Self {
        Self::new()
    }
}

File: sdk/src/intent/parser.ts
import {
  Intent,
  IntentType,
  SwapParams,
  MultiHopSwapParams,
  StakeParams,
  UnstakeParams,
  ProvideLiquidityParams,
  TransferParams,
  IntentParams,
  KNOWN_MINTS,
} from "../types";
import {
  DSL_VERBS,
  DslTokens,
  IntentValidationError,
  IntentValidationResult,
  JsonIntentInput,
} from "./types";

/**
 * Parses user intent from either a DSL string or JSON object.
 *
 * DSL format examples:
 *   "swap 100 USDC to SOL"
 *   "swap 100 USDC to SOL slippage 50"
 *   "stake 10 SOL"
 *   "unstake 5 SOL"
 *   "transfer 50 USDC to <pubkey>"
 *   "provide 100 USDC and 0.5 SOL"
 *   "swap 100 USDC to RAY to SOL" (multi-hop)
 */
export class IntentParser {
  /**
   * Parse an intent from a string (DSL) or object (JSON).
   */
  parse(input: string | object): Intent {
    if (typeof input === "string") {
      return this.parseDsl(input);
    }
    return this.parseJson(input as JsonIntentInput);
  }

  /**
   * Parse a DSL string into an Intent.
   */
  private parseDsl(input: string): Intent {
    const normalized = input.trim().toLowerCase();
    const tokens = this.tokenize(normalized);

    const intentType = DSL_VERBS[tokens.verb];
    if (!intentType) {
      throw new Error(`Unknown intent verb: '${tokens.verb}'`);
    }

    let params: IntentParams;

    switch (intentType) {
      case IntentType.Swap:
        params = this.buildSwapParams(tokens, normalized);
        break;
      case IntentType.Stake:
        params = this.buildStakeParams(tokens);
        break;
      case IntentType.Unstake:
        params = this.buildUnstakeParams(tokens);
        break;
      case IntentType.Transfer:
        params = this.buildTransferParams(tokens);
        break;
      case IntentType.ProvideLiquidity:
        params = this.buildLiquidityParams(tokens);
        break;
      default:
        throw new Error(`Unsupported intent type: ${intentType}`);
    }

    // Check for multi-hop swap (e.g., "swap 100 USDC to RAY to SOL")
    if (intentType === IntentType.Swap) {
      const multiHopParams = this.detectMultiHop(normalized, tokens);
      if (multiHopParams) {
        return {
          type: IntentType.MultiHopSwap,
          params: multiHopParams,
          id: this.generateId(),
          createdAt: Date.now(),
          priority: 1,
        };
      }
    }

    return {
      type: intentType,
      params,
      id: this.generateId(),
      createdAt: Date.now(),
      priority: 1,
    };
  }

  /**
   * Parse a JSON object into an Intent.
   */
  private parseJson(input: JsonIntentInput): Intent {
    const type = this.resolveIntentType(input.type);
    const params = this.buildParamsFromJson(type, input.params);

    return {
      type,
      params,
      id: this.generateId(),
      createdAt: Date.now(),
      priority: input.priority ?? 1,
    };
  }

  /**
   * Tokenize a DSL string.
   */
  private tokenize(input: string): DslTokens {
    const words = input.split(/\s+/);
    if (words.length < 2) {
      throw new Error("Intent string too short; expected at least a verb and amount");
    }

    const verb = words[0];
    const amountStr = words[1];
    const amount = parseFloat(amountStr);
    if (isNaN(amount)) {
      throw new Error(`Invalid amount: '${amountStr}'`);
    }

    const tokenA = words[2]?.toUpperCase() ?? "";
    const preposition = words[3] ?? "";
    const tokenB = words[4]?.toUpperCase() ?? "";

    // Parse extras (key-value pairs after main tokens)
    const extras: Record<string, string> = {};
    for (let i = 5; i < words.length; i += 2) {
      if (i + 1 < words.length) {
        extras[words[i]] = words[i + 1];
      }
    }

    // Check for "and" preposition for liquidity
    let amountB: number | undefined;
    if (preposition === "and" && tokenB) {
      const amountBStr = words[4];
      amountB = parseFloat(amountBStr ?? "0");
      if (isNaN(amountB)) amountB = undefined;
    }

    return { verb, amount, tokenA, preposition, tokenB, amountB, extras };
  }

  /**
   * Build swap parameters from DSL tokens.
   */
  private buildSwapParams(tokens: DslTokens, raw: string): SwapParams {
    const inputMint = this.resolveMint(tokens.tokenA);
    const outputMint = this.resolveMint(tokens.tokenB);
    const slippageBps = tokens.extras["slippage"]
      ? parseInt(tokens.extras["slippage"], 10)
      : 50;

    return {
      inputMint,
      outputMint,
      amount: tokens.amount,
      slippageBps,
    };
  }

  /**
   * Build stake parameters from DSL tokens.
   */
  private buildStakeParams(tokens: DslTokens): StakeParams {
    return {
      amount: tokens.amount,
      validatorVote: tokens.extras["validator"],
    };
  }

  /**
   * Build unstake parameters from DSL tokens.
   */
  private buildUnstakeParams(tokens: DslTokens): UnstakeParams {
    return {
      amount: tokens.amount,
      stakeAccount: tokens.extras["account"],
    };
  }

  /**
   * Build transfer parameters from DSL tokens.
   */
  private buildTransferParams(tokens: DslTokens): TransferParams {
    const mint = this.resolveMint(tokens.tokenA);
    return {
      mint,
      amount: tokens.amount,
      recipient: tokens.tokenB || tokens.extras["to"] || "",
    };
  }

  /**
   * Build liquidity parameters from DSL tokens.
   */
  private buildLiquidityParams(tokens: DslTokens): ProvideLiquidityParams {
    // "provide 100 USDC and 0.5 SOL"
    const words = tokens.verb === "provide" ? tokens : tokens;
    const tokenAMint = this.resolveMint(tokens.tokenA);

    // Re-parse for the "and X TOKEN" pattern
    const tokenBMint = tokens.extras["tokenb"]
      ? this.resolveMint(tokens.extras["tokenb"])
      : this.resolveMint(tokens.tokenB);
    const amountB = tokens.amountB ?? 0;

    return {
      tokenAMint,
      tokenBMint,
      amountA: tokens.amount,
      amountB,
      poolAddress: tokens.extras["pool"],
    };
  }

  /**
   * Detect multi-hop swaps: "swap 100 USDC to RAY to SOL"
   */
  private detectMultiHop(
    raw: string,
    tokens: DslTokens
  ): MultiHopSwapParams | null {
    const words = raw.split(/\s+/);
    // Find all "to" prepositions after the verb
    const toIndices: number[] = [];
    for (let i = 3; i < words.length; i++) {
      if (words[i] === "to") toIndices.push(i);
    }

    if (toIndices.length < 2) return null;

    // Build hops
    const hops: Array<{ inputMint: string; outputMint: string }> = [];
    let currentInput = tokens.tokenA;

    for (const idx of toIndices) {
      const outputToken = words[idx + 1]?.toUpperCase() ?? "";
      if (!outputToken) continue;
      hops.push({
        inputMint: this.resolveMint(currentInput),
        outputMint: this.resolveMint(outputToken),
      });
      currentInput = outputToken;
    }

    if (hops.length < 2) return null;

    const slippageBps = tokens.extras["slippage"]
      ? parseInt(tokens.extras["slippage"], 10)
      : 50;

    return {
      hops,
      amount: tokens.amount,
      slippageBps,
    };
  }

  /**
   * Resolve a token symbol to its mint address.
   */
  private resolveMint(symbol: string): string {
    const upper = symbol.toUpperCase();
    return KNOWN_MINTS[upper] ?? symbol;
  }

  /**
   * Resolve an intent type string to the IntentType enum.
   */
  private resolveIntentType(typeStr: string): IntentType {
    const lower = typeStr.toLowerCase().replace(/[-_\s]/g, "");
    const mapping: Record<string, IntentType> = {
      swap: IntentType.Swap,
      multihopswap: IntentType.MultiHopSwap,
      multi_hop_swap: IntentType.MultiHopSwap,
      stake: IntentType.Stake,
      unstake: IntentType.Unstake,
      provideliquidity: IntentType.ProvideLiquidity,
      provide_liquidity: IntentType.ProvideLiquidity,
      addliquidity: IntentType.ProvideLiquidity,
      transfer: IntentType.Transfer,
      send: IntentType.Transfer,
    };
    const result = mapping[lower];
    if (!result) throw new Error(`Unknown intent type: '${typeStr}'`);
    return result;
  }

  /**
   * Build IntentParams from JSON input based on type.
   */
  private buildParamsFromJson(
    type: IntentType,
    raw: Record<string, unknown>
  ): IntentParams {
    switch (type) {
      case IntentType.Swap:
        return {
          inputMint: this.resolveMint(String(raw.inputMint ?? raw.from ?? "")),
          outputMint: this.resolveMint(String(raw.outputMint ?? raw.to ?? "")),
          amount: Number(raw.amount ?? 0),
          slippageBps: Number(raw.slippageBps ?? raw.slippage ?? 50),
          maxAccounts: raw.maxAccounts ? Number(raw.maxAccounts) : undefined,
        } as SwapParams;

      case IntentType.MultiHopSwap: {
        const hops = (raw.hops as Array<Record<string, string>>) ?? [];
        return {
          hops: hops.map((h) => ({
            inputMint: this.resolveMint(String(h.inputMint ?? h.from ?? "")),
            outputMint: this.resolveMint(String(h.outputMint ?? h.to ?? "")),
          })),
          amount: Number(raw.amount ?? 0),
          slippageBps: Number(raw.slippageBps ?? 50),
        } as MultiHopSwapParams;
      }

      case IntentType.Stake:
        return {
          amount: Number(raw.amount ?? 0),
          validatorVote: raw.validatorVote
            ? String(raw.validatorVote)
            : undefined,
        } as StakeParams;

      case IntentType.Unstake:
        return {
          amount: Number(raw.amount ?? 0),
          stakeAccount: raw.stakeAccount
            ? String(raw.stakeAccount)
            : undefined,
        } as UnstakeParams;

      case IntentType.ProvideLiquidity:
        return {
          tokenAMint: this.resolveMint(String(raw.tokenAMint ?? raw.tokenA ?? "")),
          tokenBMint: this.resolveMint(String(raw.tokenBMint ?? raw.tokenB ?? "")),
          amountA: Number(raw.amountA ?? 0),
          amountB: Number(raw.amountB ?? 0),
          poolAddress: raw.poolAddress ? String(raw.poolAddress) : undefined,
        } as ProvideLiquidityParams;

      case IntentType.Transfer:
        return {
          mint: this.resolveMint(String(raw.mint ?? raw.token ?? "")),
          amount: Number(raw.amount ?? 0),
          recipient: String(raw.recipient ?? raw.to ?? ""),
        } as TransferParams;

      default:
        throw new Error(`Cannot build params for type: ${type}`);
    }
  }

  /**
   * Generate a unique intent ID.
   */
  private generateId(): string {
    return `intent_${Date.now()}_${Math.random().toString(36).slice(2, 10)}`;
  }
}

/**
 * Validates parsed intents for correctness.
 */
export class IntentValidator {
  /**
   * Validate an intent.
   */
  validate(intent: Intent): IntentValidationResult {
    const errors: IntentValidationError[] = [];
    const warnings: string[] = [];

    // Common validations
    if (!intent.id) {
      errors.push({ field: "id", message: "Intent ID is required" });
    }
    if (!intent.type) {
      errors.push({ field: "type", message: "Intent type is required" });
    }

    switch (intent.type) {
      case IntentType.Swap:
        this.validateSwap(intent.params as SwapParams, errors, warnings);
        break;
      case IntentType.MultiHopSwap:
        this.validateMultiHopSwap(
          intent.params as MultiHopSwapParams,
          errors,
          warnings
        );
        break;
      case IntentType.Stake:
        this.validateStake(intent.params as StakeParams, errors, warnings);
        break;
      case IntentType.Unstake:
        this.validateUnstake(intent.params as UnstakeParams, errors, warnings);
        break;
      case IntentType.ProvideLiquidity:
        this.validateLiquidity(
          intent.params as ProvideLiquidityParams,
          errors,
          warnings
        );
        break;
      case IntentType.Transfer:
        this.validateTransfer(
          intent.params as TransferParams,
          errors,
          warnings
        );
        break;
    }

    return {
      valid: errors.length === 0,
      errors,
      warnings,
    };
  }

  private validateSwap(
    params: SwapParams,
    errors: IntentValidationError[],
    warnings: string[]
  ): void {
    if (!params.inputMint) {
      errors.push({
        field: "inputMint",
        message: "Input mint is required",
      });
    }
    if (!params.outputMint) {
      errors.push({
        field: "outputMint",
        message: "Output mint is required",
      });
    }
    if (params.inputMint === params.outputMint) {
      errors.push({
        field: "outputMint",
        message: "Input and output mints must be different",
        value: params.outputMint,
      });
    }
    if (params.amount <= 0) {
      errors.push({
        field: "amount",
        message: "Amount must be greater than 0",
        value: params.amount,
      });
    }
    if (params.slippageBps < 0 || params.slippageBps > 10_000) {
      errors.push({
        field: "slippageBps",
        message: "Slippage must be between 0 and 10000 bps",
        value: params.slippageBps,
      });
    }
    if (params.slippageBps > 500) {
      warnings.push(
        `High slippage tolerance: ${params.slippageBps} bps (${params.slippageBps / 100}%)`
      );
    }
    this.validateMintAddress(params.inputMint, "inputMint", errors);
    this.validateMintAddress(params.outputMint, "outputMint", errors);
  }

  private validateMultiHopSwap(
    params: MultiHopSwapParams,
    errors: IntentValidationError[],
    warnings: string[]
  ): void {
    if (!params.hops || params.hops.length < 2) {
      errors.push({
        field: "hops",
        message: "Multi-hop swap requires at least 2 hops",
      });
    }
    if (params.amount <= 0) {
      errors.push({
        field: "amount",
        message: "Amount must be greater than 0",
        value: params.amount,
      });
    }
    // Validate hop chain continuity
    if (params.hops && params.hops.length >= 2) {
      for (let i = 0; i < params.hops.length - 1; i++) {
        if (params.hops[i].outputMint !== params.hops[i + 1].inputMint) {
          errors.push({
            field: `hops[${i}]`,
            message: `Hop chain broken: output of hop ${i} doesn't match input of hop ${i + 1}`,
          });
        }
      }
    }
    if (params.hops && params.hops.length > 4) {
      warnings.push(
        `${params.hops.length} hops may result in high slippage and fees`
      );
    }
  }

  private validateStake(
    params: StakeParams,
    errors: IntentValidationError[],
    _warnings: string[]
  ): void {
    if (params.amount <= 0) {
      errors.push({
        field: "amount",
        message: "Stake amount must be greater than 0",
        value: params.amount,
      });
    }
    if (params.amount < 0.01) {
      errors.push({
        field: "amount",
        message: "Minimum stake is 0.01 SOL",
        value: params.amount,
      });
    }
  }

  private validateUnstake(
    params: UnstakeParams,
    errors: IntentValidationError[],
    _warnings: string[]
  ): void {
    if (params.amount <= 0) {
      errors.push({
        field: "amount",
        message: "Unstake amount must be greater than 0",
        value: params.amount,
      });
    }
  }

  private validateLiquidity(
    params: ProvideLiquidityParams,
    errors: IntentValidationError[],
    _warnings: string[]
  ): void {
    if (!params.tokenAMint) {
      errors.push({ field: "tokenAMint", message: "Token A mint is required" });
    }
    if (!params.tokenBMint) {
      errors.push({ field: "tokenBMint", message: "Token B mint is required" });
    }
    if (params.amountA <= 0) {
      errors.push({
        field: "amountA",
        message: "Amount A must be greater than 0",
      });
    }
    if (params.amountB <= 0) {
      errors.push({
        field: "amountB",
        message: "Amount B must be greater than 0",
      });
    }
  }

  private validateTransfer(
    params: TransferParams,
    errors: IntentValidationError[],
    _warnings: string[]
  ): void {
    if (!params.mint) {
      errors.push({ field: "mint", message: "Token mint is required" });
    }
    if (params.amount <= 0) {
      errors.push({
        field: "amount",
        message: "Transfer amount must be greater than 0",
      });
    }
    if (!params.recipient) {
      errors.push({
        field: "recipient",
        message: "Recipient address is required",
      });
    }
    this.validateMintAddress(params.recipient, "recipient", errors);
  }

  /**
   * Basic validation that a string looks like a valid base58 Solana address.
   */
  private validateMintAddress(
    address: string,
    field: string,
    errors: IntentValidationError[]
  ): void {
    if (!address) return;
    // Base58 characters only, 32-44 chars
    if (!/^[1-9A-HJ-NP-Za-km-z]{32,44}$/.test(address)) {
      errors.push({
        field,
        message: `Invalid Solana address format: '${address}'`,
        value: address,
      });
    }
  }
}
